cranelift-codegen/meta/src/isa/x86/recipes.rs

```diff
@@ -3035,10 +3036,87 @@ pub(crate) fn define<'shared>(
     recipes.add_recipe(
         EncodingRecipeBuilder::new("safepoint", &formats.multiary, 0).emit(
             r#"
-            sink.add_stackmap(args, func, isa);
-        "#,
+                sink.add_stackmap(args, func, isa);
+            "#,
         ),
     );
 
+    /*recipes.add_recipe(
+        EncodingRecipeBuilder::new("tls_value", &formats.unary_global_value, 19)
+            .operands_out(vec![gpr])
+            .emit(
+                r#"
+                    // output %rax
+                    // clobbers %rdi
+
+                    // leaq @tlsld(%rip), %rdi
+                    sink.put1(0x48);
+                    sink.put1(0x8d);
+                    sink.put1(0x3d);
+                    sink.reloc_external(Reloc::ElfX86_64TlsLd,
+                                        &func.global_values[global_value].symbol_name(),
+                                        -4);
+                    sink.put4(0);
+
+                    // callq __tls_get_addr-4
+                    sink.put1(0xe8);
+                    sink.reloc_external(Reloc::X86CallPLTRel4,
+                                        &ExternalName::LibCall(LibCall::ElfTlsGetAddr),
+                                        -4);
+                    sink.put4(0);
+
+                    // leaq @dtpoff32(%rax), %rax
+                    sink.put1(0x48);
+                    sink.put1(0x8d);
+                    sink.put1(0x80);
+                    sink.reloc_external(Reloc::ElfX86_64DtpOff32,
+                                        &func.global_values[global_value].symbol_name(),
+                                        0);
+                    sink.put4(0);
+                "#,
+            ),
+    );*/
+
+    recipes.add_recipe(
+        EncodingRecipeBuilder::new("elf_tlsld", &formats.unary_global_value, 7)
+            // FIXME Should use gpr instead of reg_rdi, but that causes an unnecessary regmove instruction.
+            .operands_out(vec![reg_rdi])
+            .emit(
+                r#"
+                    // leaq @tlsld(%rip), %rdi
+                    sink.put1(0b0100_1000);
+                    const LEA: u8 = 0x8d;
```

Comment:
This could probably be imported from `opcodes.rs`

Reply:
Do you know what the most ideomatic way would be to encode this? Using a Template instead of Recipe and then putting the rex.w prefix and opcode in encodings.rs? And if so, what should be the declaration of that Template?

Reply:
I just meant using the `LEA` already in https://github.com/CraneStation/cranelift/blob/a18dec8b22739e95a3f0395de26e5539b7e04afb/cranelift-codegen/meta/src/isa/x86/opcodes.rs#L132

But I think I see what you mean: we've been able to keep most of the other opcodes in `encodings.rs` and use templates, a la https://github.com/CraneStation/cranelift/blob/cabbacbe63477fcc7d1f1e36717d84f08f749a98/cranelift-codegen/meta/src/isa/x86/recipes.rs#L2417-L2428

I think that `{{PUT_OP}} ...` stuff should work for you here too. (But our lack of clarity on how to use this is probably a motivator for taking another look at recipes/templates, #955).

Reply:
Now that I switched to the TLS GD model, I dont even think it would be possible to use `Template` anymore, as I need a specfic sequence of inst prefixes and two asm insts, so using two clif insts may cause problems if some pass decides to put an inst inbetween.